cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStoreProvider.java

```diff
@@ -0,0 +1,543 @@
+
+  /*
+   A cache for metadata.
+   The assumption is that the number of entries won't be large enough to worry about memory.
+  */
+  private final Map<String, SecureStoreMetadata> cache = new HashMap<>();
+  private final char[] password;
+  private final Path path;
+  private final Lock readLock;
+  private final Lock writeLock;
+  private final CConfiguration cConf;
+
+  private KeyStore keyStore;
+  private boolean changed = false;
+
+  private FileSecureStoreProvider(CConfiguration cConf) {
+    this.cConf = cConf;
+    // Get the path to the keystore file
+    String pathString = cConf.get(Constants.Security.Store.SECURE_STORE_FILE_PATH);
+    if (pathString == null || pathString.isEmpty()) {
+      pathString = SECURE_STORE_DEFAULT_FILE_PATH;
+    }
+    Path dir = Paths.get(pathString);
+    path = dir.resolve(SECURE_STORE_DEFAULT_FILE_NAME);
+
+    // Get the keystore password
+    String passwordString = cConf.get(Constants.Security.Store.SECURE_STORE_FILE_PASSWORD);
+    if (passwordString == null || passwordString.isEmpty()) {
+      password = SECURE_STORE_DEFAULT_PASSWORD;
+    } else {
+      password = passwordString.toCharArray();
+    }
+
+    ReadWriteLock lock = new ReentrantReadWriteLock(true);
+    readLock = lock.readLock();
+    writeLock = lock.writeLock();
+  }
+
+  public static FileSecureStoreProvider getInstance(CConfiguration cConf) throws IOException {
+    FileSecureStoreProvider fileSecureStoreProvider = new FileSecureStoreProvider(cConf);
+    // Locate and load the keystore
+    fileSecureStoreProvider.locateKeystore();
+    return fileSecureStoreProvider;
+  }
+
+  /**
+   * Stores an element in the secure store. If the key already exists then delete it first.
+   * @param name
+   * @param data
+   * @param properties
+   * @throws IOException
+   */
+  void put(String name, byte[] data, Map<String, String> properties) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    writeLock.lock();
+    try {
+      if (keyStore.containsAlias(name) || cache.containsKey(metaKey)) {
+        // Clear the existing key so that we can write the new one.
+        delete(name);
+      }
+      keyStore.setKeyEntry(name, new SecretKeySpec(data, ALGORITHM_PROXY),
+                           password, null);
+      SecureStoreMetadata meta = FileSecureStoreMetadata.of(name, properties);
+      /*
+        The data was written successfully to the key store. Now try to write the metddata.
+        If this fails then try to delete the data and throw an exception.
+       */
+      try {
+        keyStore.setKeyEntry(metaKey, new KeyMetadata(meta), password, null);
+      } catch (KeyStoreException k) {
+        keyStore.deleteEntry(name);
+        throw k;
+      }
+      cache.put(metaKey, meta);
+    } catch (KeyStoreException e) {
+      throw new IOException("Failed to store the key. ", e);
+    } finally {
+      writeLock.unlock();
+    }
+    // Attempt to persist the store.
```

Comment:
Not necessarily. The data is still in memory and can be served. It could be written successfully in the next flush attempt. We can choose to fail the put on a flush failure but it is not mandatory.